geo/src/algorithm/outlier_detection.rs

```diff
@@ -0,0 +1,417 @@
+impl<T> Display for LofScore<T>
+where
+    T: std::fmt::Debug,
+{
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(
+            f,
+            "Point index: {}, LOF Score: {:?}",
+            self.index, self.score
+        )
+    }
+}
+
+/// This struct allows multiple detection operations to be run on a point set using varying `k_neighbours` sizes
+/// without having to rebuild the underlying spatial index. Its [`PreparedDetector::outliers`] method
+/// has the same signature as [`OutlierDetection::outliers`], but retains the underlying spatial index and point set
+/// for greater efficiency.
+#[derive(Clone, Debug)]
+pub struct PreparedDetector<'a, T>
+where
+    T: GeoFloat,
+{
+    tree: RTree<GeomWithData<Point<T>, usize>>,
+    points: &'a [Point<T>],
+}
+
+impl<'a, T> PreparedDetector<'a, T>
+where
+    T: GeoFloat + Sum,
+{
+    /// Create a new "prepared" detector which allows repeated LOF algorithm calls with varying neighbour sizes
+    fn new(points: &'a [Point<T>]) -> Self {
+        let geoms: Vec<GeomWithData<_, usize>> = points
+            .iter()
+            .enumerate()
+            .map(|(idx, point)| GeomWithData::new(*point, idx))
+            .collect();
+        let tree = RTree::bulk_load(geoms);
+        Self { tree, points }
+    }
+
+    /// The LOF algorithm. `k_neighbours` specifies the number of neighbours to use for local outlier
+    /// classification. The paper linked above (see p. 100) suggests a `k_neighbours` value of 10 - 20
+    /// as a lower bound for "real-world"
+    /// data. The accuracy of the algorithm can be increased (see p. 101) by performing multiple passes
+    /// over the data using a range of `k_neighbours` values, then aggregating these, taking e.g.
+    /// the minimum or the mean.
+    ///
+    /// If `k_neighbours` >= points in the set, input points will be returned with a `score` of 1.
+    /// If there are at least `k_neighbours` duplicate points of an input point, LOF scores can be ∞ or `NaN`.
+    /// It is thus advisable to **deduplicate** or otherwise ensure the uniqueness of the input points.
+    ///
+    /// # Examples
+    ///
+    /// `MultiPoint`:
+    ///
+    /// ```
+    /// use approx::assert_relative_eq;
+    /// use geo::OutlierDetection;
+    /// use geo::{point, MultiPoint};
+    ///
+    /// let v = vec![
+    ///     point!(x: 0.0, y: 0.0),
+    ///     point!(x: 0.0, y: 1.0),
+    ///     point!(x: 3.0, y: 0.0),
+    ///     point!(x: 1.0, y: 1.0),
+    /// ];
+    /// let mp = MultiPoint(v);
+    ///
+    /// let lofscores = mp.outliers(3);
+    ///
+    /// assert_relative_eq!(lofscores[2].score, 3.3333333333333335);
+    /// ```
+    pub fn outliers(&self, kneighbours: usize) -> Vec<LofScore<T>> {
+        lof(self.points, &self.tree, kneighbours)
+    }
+}
+
+fn lof<T>(
+    points: &[Point<T>],
+    tree: &RTree<GeomWithData<Point<T>, usize>>,
+    kneighbours: usize,
+) -> Vec<LofScore<T>>
+where
+    T: GeoFloat + Sum,
+{
```

Comment:
```suggestion
{
    assert!(kneighbours > 0);
```

Would give a better panic message than the later `unwrap` panic.

Also, I'm not sure the meaning of this algo when kneighbors == 1, but maybe that's fine and not worth panicking over.

Reply:
1 will just give lower-quality results. Note that passing `kneighbours = 0` won't cause a panic down the line - we immediately return all the LOF scores set to 1, which is essentially saying "we couldn't calculate them". I've added a `debug_assert` – dunno if that's good practice though.

Reply:
Hmm, when I run this:
```
        let v = [
            Point::new(0.0, 1.0),
            Point::new(0.0, 2.0),
            Point::new(3.0, 1.0),
            Point::new(1.0, 1.0),
        ];

        let scores = v.outliers(0);
```

I see this panic:
```
called `Option::unwrap()` on a `None` value
thread 'algorithm::outlier_detection::tests::test_same_points' panicked at 'called `Option::unwrap()` on a `None` value', geo/src/algorithm/outlier_detection.rs:232:18
stack backtrace:
   0: rust_begin_unwind
             at /rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/std/src/panicking.rs:584:5
   1: core::panicking::panic_fmt
             at /rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/core/src/panicking.rs:142:14
   2: core::panicking::panic
             at /rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/core/src/panicking.rs:48:5
   3: core::option::Option<T>::unwrap
             at /rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/core/src/option.rs:775:21
   4: geo::algorithm::outlier_detection::lof::{{closure}}
             at ./src/algorithm/outlier_detection.rs:228:28
   5: core::iter::adapters::map::map_fold::{{closure}}
             at /rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/core/src/iter/adapters/map.rs:84:28
   6: core::iter::traits::iterator::Iterator::fold
             at /rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/core/src/iter/traits/iterator.rs:2370:21
   7: <core::iter::adapters::map::Map<I,F> as core::iter::traits::iterator::Iterator>::fold
             at /rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/core/src/iter/adapters/map.rs:124:9
   8: core::iter::traits::iterator::Iterator::for_each
             at /rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/core/src/iter/traits/iterator.rs:787:9
   9: <alloc::vec::Vec<T,A> as alloc::vec::spec_extend::SpecExtend<T,I>>::spec_extend
             at /rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/alloc/src/vec/spec_extend.rs:40:17
  10: <alloc::vec::Vec<T> as alloc::vec::spec_from_iter_nested::SpecFromIterNested<T,I>>::from_iter
             at /rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/alloc/src/vec/spec_from_iter_nested.rs:62:9
  11: <alloc::vec::Vec<T> as alloc::vec::spec_from_iter::SpecFromIter<T,I>>::from_iter
             at /rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/alloc/src/vec/spec_from_iter.rs:33:9
  12: <alloc::vec::Vec<T> as core::iter::traits::collect::FromIterator<T>>::from_iter
             at /rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/alloc/src/vec/mod.rs:2645:9
  13: core::iter::traits::iterator::Iterator::collect
             at /rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/core/src/iter/traits/iterator.rs:1792:9
  14: geo::algorithm::outlier_detection::lof
             at ./src/algorithm/outlier_detection.rs:224:48
  15: geo::algorithm::outlier_detection::PreparedDetector<T>::outliers
             at ./src/algorithm/outlier_detection.rs:192:9
  16: <[geo_types::geometry::point::Point<T>] as geo::algorithm::outlier_detection::OutlierDetection<T>>::outliers
             at ./src/algorithm/outlier_detection.rs:299:9
  17: geo::algorithm::outlier_detection::tests::test_same_points
             at ./src/algorithm/outlier_detection.rs:429:22
  18: geo::algorithm::outlier_detection::tests::test_same_points::{{closure}}
             at ./src/algorithm/outlier_detection.rs:420:5
  19: core::ops::function::FnOnce::call_once
             at /rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/core/src/ops/function.rs:248:5
  20: core::ops::function::FnOnce::call_once
             at /rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/core/src/ops/function.rs:248:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

Reply:
Ugh yes my fault: we only check that neighbours < points. But we could just extend that check to neighbours > 0 instead of just panicking? Which is better: indeterminate result, or a panic?

Reply:
Neighbors less than 0 seems almost certainly to be a programmer error, so I'm very biased for panicing in debug builds.

Doing our best to avoid panics in release builds seems nice though, so I have a slight bias in favor of "blundering on" there.

Reply:
We now panic in release, and return 1 in release if `neighbours > points` OR `neighbours < 1`.

Reply:
> We now panic in release,

We now panic in ~~release~~ debug,

I think this was just a typo in your github comment. The code looks right -thank you for making this change!